### DISCUSSION TITLE
cannot refresh token.. (but everything else works)

### DISCUSSION THREAD

**Discussion by LTa2022** [etsy/open-api#864] [Q&A]
Hi guys, 
i got basically everything working.. except  .... getting a refresh token. I use rest sharp. I have tried approaches from different sources.. 
The token definitely works as i am using it elsewhere ..
I tried:  
different headers but as i understand none are necessary. 
add jsonbody and andbody with an anonymous object
addparameter
and another approach i found using  dictionary and httpclient.

Everything I ever get is   {"error":"invalid_grant","error_description":"refresh_token is invalid"}

Thanks!

i got basically everything working.. except  .... getting a refresh token. I use rest sharp. I have tried approaches from different sources.. 
The token definitely works as i am using it elsewhere ..
I tried:  
different headers but as i understand none are necessary. 
add jsonbody and andbody with an anonymous object
addparameter with post and 

   public static  RestResponse RefreshToken(string token)
        {

            var request = new RestRequest(@$"public/oauth/token", Method.Post);
            //request.AddHeader("x-api-key", keystring);
            //request.AddHeader("Content-Type", "application/x-www-form-urlencoded");

            //request.AddHeader("content-type", "application/x-www-form-urlencoded");

            //request.AddJsonBody(new
            //{
            //    grant_type = "refresh_token",
            //    client_id = keystring,
            //    refresh_token = token

            //});

            //request.AddParameter("grant_type", "refresh_token", ParameterType.GetOrPost);
            //request.AddParameter("refresh_token", token, ParameterType.GetOrPost);
            //request.AddParameter("client_id", keystring, ParameterType.GetOrPost);

            refre();

            return apiclient.Execute(request);

        }
        public static async void refre()
        {
            string url = "https://openapi.etsy.com/v3/public/oauth/token";

            Dictionary<string, string> data = new Dictionary<string, string>
    {
        { "grant_type", "refresh_token" },
        { "client_id", keystring },
        { "refresh_token", token }
    };

            var httpClient = new HttpClient();
            var response = await httpClient.PostAsync(url, new FormUrlEncodedContent(data));

           var contents = await response.Content.ReadAsStringAsync();

        }

**Comment by turboturtleD** [comment]
Since this is all a URL encoded string, can you log the output of the POST URL?

Should look just like:

https://api.etsy.com/v3/public/oauth/token?grant_type=refresh_token&client_id=1aa2bb33c44d55eeeeee6fff&refresh_token=12345678.JNGIJtvLmwfDMhlYoOJl8aLR1BWottyHC6yhNcET-eC7RogSR5e1GTIXGrgrelWZalvh3YvvyLfKYYqvymd-u37Sjtx

**Comment by Gareth064** [comment]
I'm not using RestSharp, just the out of the box .NET HttpClient and System.Text.Json.

This is my working code

```cs
public async Task<EtsyTokenResponse> GetRefreshAccessTokenAsync(string refreshToken)
    {
        EtsyTokenResponse? token = null;

        var formData = new List<KeyValuePair<string, string>>
        {
            new KeyValuePair<string, string>("grant_type", "refresh_token"),
            new KeyValuePair<string, string>("client_id", $"{clientId}"),
            new KeyValuePair<string, string>("refresh_token", $"{refreshToken}")
        };

        HttpRequestMessage baseRequest = new(HttpMethod.Post, "https://api.etsy.com/v3/public/oauth/token");
        baseRequest.Content = new FormUrlEncodedContent(formData);

        var httpClient = httpClientFactory.CreateClient();

        var request = httpClient.SendAsync(baseRequest);

        using (var response = await request)
        {
            string responseString = await response.Content.ReadAsStringAsync();
            token = JsonSerializer.Deserialize<EtsyTokenResponse>(responseString);
        }

        return token!;
    }
```
And the EtsyTokenResponse model

```cs
public class EtsyTokenResponse
{
    [JsonPropertyName("access_token")]
    public string AccessToken { get; set; }

    [JsonPropertyName("token_type")]
    public string TokenType { get; set; }

    [JsonPropertyName("expires_in")]
    public long ExpiresIn { get; set; }

    [JsonPropertyName("refresh_token")]
    public string RefreshToken { get; set; }
}
```

**Comment by LTa2022** [comment]
Hi guys thanks for the approaches!

For the first I used: 
       public static void Refresh2()
        {

            var url = @$"https://api.etsy.com/v3/public/oauth/token?grant_type=refresh_token&client_id={keystring}&refresh_token={token}";

            var req = new RestRequest(url, Method.Post);

            RestClient c = new RestClient();

            var resp = c.Execute(req);

        }
response: 
{"error":"invalid_request","error_description":"grant_type is required"}

But as the second is a as is working code, im still getting:
{"error":"invalid_grant","error_description":"refresh_token is invalid"}

I verified again, the token works when e.g. updating a listing.

**Comment by LTa2022** [comment]
Hi, 
is there anything else I could check? 
Just a recap what my code does so far: it opens the etsy login page, where i confirm all the rights, then receive a token. 
That token works for everything i tested so far, e.g. editing, adding listings. But after 1 hour it's invalid. 
When I request a refresh token as described in last message i get 
{"error":"invalid_grant","error_description":"refresh_token is invalid"}

**Reply by turboturtleD** [reply]
Just to clarify here (maybe this is the issue), you need to request a new access token (not a refresh token).   Once the access token expires, it is no longer useful for any type of request.  Instead, the refresh token (that you received when you got the access token) must be used and exchanged for a new access token (as well as a new refresh token).  Everything else looks pretty clean, so not sure how to help if this is not the issue.

**Reply by Gareth064** [reply]
Just so we are clear, you are calling the right endpoint (based on the code you mentioned above) and are passing in the "Refresh Token", which you recieved when you got your "Access_Token" from the first call (which involved the etsy login etc.)

![image](https://user-images.githubusercontent.com/22307223/223541635-ecc141f0-e355-418b-b4fb-238344e071e0.png)
